test/adapter.js

```diff
@@ -0,0 +1,274 @@
+
+		assert.equal(parent.length, 1);
+	});
+
+	it("should sibling", function(){
+		var body = getBody(html);
+		var sibling = select("div + span", body, options);
+
+		assert.equal(sibling.length, 1);
+	});
+
+	it("should adjacent", function(){
+		var body = getBody(html);
+		var adjacent = select("strong ~ em", body, options);
+
+		assert.equal(adjacent.length, 1);
+	});
+
+	it("should attribute", function(){
+		var body = getBody(html);
+		var attribute = select("[class]", body, options);
+
+		assert.equal(attribute.length, 2);
+	});
+
+	it("should attribute =", function(){
+		var body = getBody(html);
+		var attribute = select("[class=\"apple\"]", body, options);
+
+		assert.equal(attribute.length, 1);
+	});
+
+	it("should attribute ~=", function(){
+		var body = getBody(html);
+		var attribute = select("[class~=\"potato\"]", body, options);
+
+		assert.equal(attribute.length, 1);
+	});
+
+	it("should attribute |=", function(){
+		var body = getBody(html);
+		var attribute = select("[id|=\"cheese\"]", body, options);
+
+		assert.equal(attribute.length, 1);
+	});
+
+	it("should attribute *=", function(){
+		var body = getBody(html);
+		var attribute = select("[id*=\"ee\"]", body, options);
+
+		assert.equal(attribute.length, 1);
+	});
+
+	it("should attribute ^=", function(){
+		var body = getBody(html);
+		var attribute = select("[id^=\"ch\"]", body, options);
+
+		assert.equal(attribute.length, 1);
+	});
+
+	it("should attribute $=", function(){
+		var body = getBody(html);
+		var attribute = select("[id$=\"burger\"]", body, options);
+
+		assert.equal(attribute.length, 1);
+	});
+
+	it("should attribute !=", function(){
+		var body = getBody(html);
+		var attribute = select("div[class!=\"apple\"]", body, options);
+
+		assert.equal(attribute.length, 1);
+	});
+
+	it("should :not", function(){
+		var body = getBody(html);
+		var pseudo = select("div:not( .apple )", body, options);
+
+		assert.equal(pseudo.length, 1);
+	});
+
+	it("should :contains", function(){
+		var body = getBody(html);
+		var pseudo = select(":contains(Hello)", body, options);
+
+		assert.equal(pseudo.length, 3);
+	});
+
+	it("should :icontains", function(){
+		var body = getBody(html);
+		var pseudo = select(":icontains(HELLO)", body, options);
+
+		assert.equal(pseudo.length, 3);
+	});
+
+	it("should :has", function(){
+		var body = getBody(html);
+		var pseudo = select(":has( em )", body, options);
+
+		assert.equal(pseudo.length, 2);
+	});
+
+  //TODO
+  /*
+  it( 'should :root', function(){
+    var body = getBody( html );
+    var pseudo = CSSselect( ":root", body, options );
+
+    assert.equal( pseudo.length, 1 );
+  });
+  */
+
+	it("should :empty", function(){
+		var body = getBody(html);
+		var pseudo = select("div:empty", body, options);
+
+		assert.equal(pseudo.length, 2);
+	});
+
+	it("should :first-child", function(){
+		var body = getBody(html);
+		var pseudo = select("div:first-child", body, options);
+
+		assert.equal(pseudo.length, 1);
+	});
+});
```

Comment:
This is awesome. Perhaps it would be an idea to have the browser adapter as a separate module, as it is starting to look like much more than just an experiment.